### PR TITLE
dot-multi-select tests

### PR DESCRIPTION
**Pull request by alfredo-dotcms**

### CODE REVIEW

File: projects/dotcms-field-elements/src/components/dot-select/dot-select.e2e.ts [hunk 1]
```diff
@@ -31,7 +31,7 @@ describe('dot-select', () => {
                 element = await page.find('dot-select');
                 await page.select('select', '');
                 await page.waitForChanges();
-                expect(element).toHaveClasses(dotTestUtil.class.emptyRequired);
+                expect(element).toHaveClasses([...dotTestUtil.class.emptyRequired, 'dot-required']);
```
Comment (hmoreras) [inline projects/dotcms-field-elements/src/components/dot-select/dot-select.e2e.ts]: test the `dot-required` in the required `prop` ?

File: projects/dotcms-field-elements/src/components/dot-multi-select/dot-multi-select.e2e.ts [hunk 2]
```diff
@@ -1,100 +1,337 @@
...
+        });
+
+        describe('hint', () => {
+            it('should render hint', async () => {
+                const value = 'test';
+                element.setProperty('hint', value);
+                await page.waitForChanges();
+
+                const hintElement = await dotTestUtil.getHint(page);
+                expect(hintElement.innerHTML).toBe(value);
```
Comment (fmontes) [inline projects/dotcms-field-elements/src/components/dot-multi-select/dot-multi-select.e2e.ts]:
```suggestion
                expect(hintElement.innerText).toBe(value);
```

File: projects/dotcms-field-elements/src/components/dot-multi-select/dot-multi-select.e2e.ts [hunk 3]
```diff
@@ -1,100 +1,337 @@
...
+                expect(hintElement).toBeNull();
+            });
+
+            it('should not break and not render with invalid data', async () => {
+                const wrongValue = [{ a: 1 }];
+                element.setProperty('hint', wrongValue);
+                await page.waitForChanges();
+
+                const hintElement = await dotTestUtil.getHint(page);
+                expect(hintElement).toBeFalsy();
```
Comment (fmontes) [inline projects/dotcms-field-elements/src/components/dot-multi-select/dot-multi-select.e2e.ts]: Is not `null`?

File: projects/dotcms-field-elements/src/components/dot-multi-select/dot-multi-select.e2e.ts [hunk 4]
```diff
@@ -1,100 +1,337 @@
...
+                expect(optionElements.length).toBe(3);
+            });
+
+            it('should not render options', async () => {
+                const optionElements = await getOptions(page);
+                expect(optionElements).toEqual([]);
+            });
+
+            it('should not break with invalid data', async () => {
+                const wrongValue = [{ a: 1 }];
```
Comment (fmontes) [inline projects/dotcms-field-elements/src/components/dot-multi-select/dot-multi-select.e2e.ts]: What happen if we pass something like `'a/1-b/2-c/3'`?

File: projects/dotcms-field-elements/src/components/dot-multi-select/dot-multi-select.e2e.ts [hunk 5]
```diff
@@ -1,100 +1,337 @@
...
+                expect(optionElements).toEqual([]);
+            });
+        });
+
+        describe('required & requiredMessage', () => {
+            it('should render required attribute in label and dot-required css class', async () => {
+                element.setProperty('required', true);
+                await page.waitForChanges();
+                const labelElement = await dotTestUtil.getDotLabel(page);
+                expect(element).toHaveClasses(['dot-required']);
```
Comment (fmontes) [inline projects/dotcms-field-elements/src/components/dot-multi-select/dot-multi-select.e2e.ts]: We should not care if the `dot-label` have a class, we only care that we're passing the right `required` attr.

File: projects/dotcms-field-elements/src/components/dot-multi-select/dot-multi-select.e2e.ts [hunk 6]
```diff
@@ -1,100 +1,337 @@
...
+                expect(await labelElement.getProperty('required')).toBe(true);
+            });
+
+            it('should render required error msg', async () => {
+                element.setProperty('required', true);
+                element.setProperty('requiredMessage', 'test');
+                await page.waitForChanges();
+                const errorElement = await dotTestUtil.getErrorMessage(page);
+                expect(errorElement.innerHTML).toBe('test');
```
Comment (fmontes) [inline projects/dotcms-field-elements/src/components/dot-multi-select/dot-multi-select.e2e.ts]:
```suggestion
                expect(errorElement.innerText).toBe('test');
```

File: projects/dotcms-field-elements/src/components/dot-multi-select/dot-multi-select.e2e.ts [hunk 7]
```diff
@@ -1,100 +1,337 @@
...
+
+            it('should not break and not render with invalid data', async () => {
+                const wrongValue = [{ a: 1 }];
+                element.setProperty('required', wrongValue);
+                element.setProperty('requiredMessage', wrongValue);
+
+                await page.waitForChanges();
+
+                const errorElement = await dotTestUtil.getErrorMessage(page);
+                expect(errorElement).toBeFalsy();
```
Comment (fmontes) [inline projects/dotcms-field-elements/src/components/dot-multi-select/dot-multi-select.e2e.ts]: Shouldn't be null?

File: projects/dotcms-field-elements/src/components/dot-multi-select/dot-multi-select.e2e.ts [hunk 8]
```diff
@@ -1,100 +1,337 @@
...
+
+                const optionElements = await getOptions(page);
+                expect(await optionElements[0].getProperty('selected')).toBe(false);
+                expect(await optionElements[1].getProperty('selected')).toBe(false);
+                expect(await optionElements[2].getProperty('selected')).toBe(false);
+            });
+
+            it('should not break with invalid data', async () => {
+                const wrongValue = [{ a: 1 }];
+                element.setProperty('options', 'a|1,b|2,c|3');
```
Comment (fmontes) [inline projects/dotcms-field-elements/src/components/dot-multi-select/dot-multi-select.e2e.ts]: What happen if we pass something like `'a/1-b/2-c/3'`?

Comment (fmontes) [reply projects/dotcms-field-elements/src/components/dot-multi-select/dot-multi-select.e2e.ts]: I'm not saying we should validate exactly the format (we have the doc) BUT we should make sure that we don't leave the value or anything in the field on a bad state.

File: projects/dotcms-field-elements/src/components/dot-multi-select/dot-multi-select.e2e.ts [hunk 9]
```diff
@@ -1,100 +1,337 @@
...
+            });
+        });
     });
 
-    describe('Events', () => {
+    describe('@Events', () => {
         beforeEach(async () => {
+            page = await newE2EPage();
+            await page.setContent(`
+            <dot-multi-select
```
Comment (fmontes) [inline projects/dotcms-field-elements/src/components/dot-multi-select/dot-multi-select.e2e.ts]: Do we need all the attr to tests the events?

File: projects/dotcms-field-elements/src/components/dot-multi-select/dot-multi-select.e2e.ts [hunk 10]
```diff
@@ -1,100 +1,337 @@
...
+                    value: '1'
+                });
             });
-            expect(spyValueChangeEvent).toHaveReceivedEventDetail({
-                name: 'testName',
-                value: '1,2'
+
+            it('should emit when 2 options selected', async () => {
+                await page.select('select', '1', '2');
+                expect(spyStatusChangeEvent).toHaveReceivedEventDetail({
```
Comment (fmontes) [inline projects/dotcms-field-elements/src/components/dot-multi-select/dot-multi-select.e2e.ts]: Is this expect not the same as https://github.com/dotCMS/core-web/pull/1034/files#diff-614e68f946feb689b1b0fe5327e8c518R290 this one?

File: projects/dotcms-field-elements/src/components/dot-select/dot-select.e2e.ts [hunk 11]
```diff
@@ -139,7 +141,6 @@ describe('dot-select', () => {
                 const value = 'test';
                 element.setProperty('hint', value);
                 await page.waitForChanges();
-
                 const hintElement = await dotTestUtil.getHint(page);
                 expect(hintElement.innerHTML).toBe(value);
```
Comment (hmoreras) [inline projects/dotcms-field-elements/src/components/dot-select/dot-select.e2e.ts]: innerText

File: projects/dotcms-field-elements/src/components/dot-multi-select/dot-multi-select.e2e.ts [hunk 12]
```diff
@@ -1,100 +1,350 @@
...
+            });
+        });
+
+        describe('required & requiredMessage', () => {
+            it('should render required error msg', async () => {
+                element.setProperty('required', true);
+                element.setProperty('requiredMessage', 'test');
+                await page.waitForChanges();
+                const errorElement = await dotTestUtil.getErrorMessage(page);
+                expect(errorElement.innerText).toBe('test');
```
Comment (fmontes) [inline projects/dotcms-field-elements/src/components/dot-multi-select/dot-multi-select.e2e.ts]: We need this tests but for the default message.

File: projects/dotcms-field-elements/src/components/dot-multi-select/dot-multi-select.e2e.ts [hunk 13]
```diff
@@ -1,100 +1,350 @@
...
+            it('should render option as selected', async () => {
+                element.setProperty('options', 'a|1,b|2');
+                element.setProperty('value', '2');
+                await page.waitForChanges();
+                const optionElements = await getOptions(page);
+                expect(await optionElements[1].getProperty('selected')).toBe(true);
+            });
+
+            it('should render options with no option selected (component\'s default behaviour)', async () => {
+                element.setProperty('options', 'a|1,b|2,c|3');
```
Comment (fmontes) [inline projects/dotcms-field-elements/src/components/dot-multi-select/dot-multi-select.e2e.ts]: Don't we need a tests to pass the option bad formatting?

File: projects/dotcms-field-elements/src/components/dot-multi-select/dot-multi-select.e2e.ts [hunk 14]
```diff
@@ -1,100 +1,350 @@
...
-            expect(spyStatusChangeEvent).toHaveReceivedEventDetail({
-                name: 'testName',
-                status: {
-                    dotPristine: true,
-                    dotTouched: false,
-                    dotValid: false
-                }
-            });
-            expect(spyValueChangeEvent).toHaveReceivedEventDetail({ name: 'testName', value: '' });
+            it('should emit on Reset', async () => {
```
Comment (fmontes) [inline projects/dotcms-field-elements/src/components/dot-multi-select/dot-multi-select.e2e.ts]: Dont we need a describe `@Method` to test the reset method?

File: projects/dotcms-field-elements/src/components/dot-multi-select/dot-multi-select.e2e.ts [hunk 15]
```diff
@@ -1,100 +1,350 @@
...
+        });
 
-    it('renders', async () => {
-        // tslint:disable-next-line:max-line-length
-        const expectedMarkup = `<dot-multi-select name=\"testName\" label=\"testLabel\" hint=\"testHint\" options=\"|,valueA|1,valueB|2\" value=\"2\" required-message=\"testErrorMsg\" required=\"true\" size=\"3\" class=\"dot-valid dot-pristine dot-untouched dot-required hydrated\"><div class=\"dot-field__label\"><label for=\"dot-testName\">testLabel</label><span class=\"dot-field__required-mark\">*</span></div><select multiple=\"\" size=\"3\" id=\"dot-testName\"><option value=\"\"></option><option value=\"1\">valueA</option><option value=\"2\">valueB</option></select><span class=\"dot-field__hint\">testHint</span></dot-multi-select>`;
-        const hint = await element.find('.dot-field__hint');
-        expect(element.outerHTML).toBe(expectedMarkup);
-        expect(hint).toBeTruthy();
-    });
+        describe('CSS classes', () => {
```
[the remark this answers is not in the document]
Comment (alfredo-dotcms) [reply projects/dotcms-field-elements/src/components/dot-multi-select/dot-multi-select.e2e.ts]: it's there 
https://github.com/dotCMS/core-web/blob/35109ab266587cacd2068c834ad8c3708a2c8272/projects/dotcms-field-elements/src/components/dot-multi-select/dot-multi-select.e2e.ts#L32

**Pull request merged**